File: rxjava/src/main/java/com/asemlab/samples/rxjava/di/NetworkModule.java
package com.asemlab.samples.rxjava.di;

import com.asemlab.samples.rxjava.remote.CountryService;
import com.asemlab.samples.rxjava.remote.HttpRoutes;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.components.SingletonComponent;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

@InstallIn(SingletonComponent.class)
@Module
public class NetworkModule {

    @Provides
    @Singleton
    public HttpLoggingInterceptor provideInterceptor() {
        return new HttpLoggingInterceptor(HttpLoggingInterceptor.Logger.DEFAULT)
                .setLevel(HttpLoggingInterceptor.Level.HEADERS);
    }

    @Provides
    @Singleton
    public OkHttpClient provideOkHttpClient(HttpLoggingInterceptor interceptor) {
        OkHttpClient client = new OkHttpClient.Builder()
                .addNetworkInterceptor(interceptor)
                .build();

        return client;
    }

    @Provides
    @Singleton
    public Retrofit provideRetrofit(OkHttpClient okhttp) {
        return new Retrofit.Builder().baseUrl(HttpRoutes.BASE_URL)
                // TODO Adding a call adapter which for creating observables
                .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .client(okhttp)
                .build();
    }

    @Provides
    @Singleton
    public CountryService provideCountryService(Retrofit retrofit) {
        return retrofit.create(CountryService.class);
    }

}

File: feature/home/src/main/java/com/asemlab/samples/feature/home/di/NetworkModule.kt
package com.asemlab.samples.feature.home.di

import com.asemlab.samples.feature.home.remote.CountryService
import com.asemlab.samples.feature.home.remote.HttpRoutes
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
class NetworkModule {


    @Singleton
    @Provides
    fun provideOkHttpClient(): OkHttpClient {
        return OkHttpClient.Builder()
            .build()
    }

    @Singleton
    @Provides
    fun provideCountryRetrofit(client: OkHttpClient): Retrofit {
        return Retrofit.Builder().baseUrl(HttpRoutes.BASE_URL)
            .addConverterFactory(GsonConverterFactory.create()).client(client).build()

    }

    @Singleton
    @Provides
    fun provideCountryService(retrofit: Retrofit): CountryService =
        retrofit.create(CountryService::class.java)

}